### ISSUE TITLE
Task Execution Strategy

### ISSUE THREAD

**Issue by pago** [pago/use-concurrent-state#3]
# Status Quo
We currently have a way to define an execution strategy for tasks:

```ts
    function* asyncTask() {
      yield d.signal;
      return 'solved';
    }
    asyncTask.strategy = {
      compose(oldTask: Task<string>, newTask: Task<string>) {
        oldTask.cancel();
        return newTask;
      },
    };
```

This API is insufficient and incomplete. For example, one major issue is that the `newTask` is already running. Therefore it is impossible to implement cases where we wait for `oldTask` to complete before starting `newTask`.

# New API

The new API should be able to handle the following cases:
- [ ] Wait for old task to complete before starting new task
- [ ] Cancel old task and start new task
- [ ] Race old task with new Task and cancel old task if new task completes first (otherwise complete old task then new task)
- [ ] Run up to `n` concurrent tasks (requires a `useTaskState` version that works with multiple concurrent tasks)
- [ ] It should be possible to specify a different strategy for every task

Overall it should be capable of expressing everything that Ember Concurrency is able to do.

## Sources of inspiration
- http://ember-concurrency.com/docs/task-concurrency
- https://folktale.origamitower.com/api/v2.3.0/en/folktale.concurrency.task.html#combining-tasks-concurrently
